File: cpp/prime-factors/prime_factors.cpp
#include "prime_factors.h"
#include <vector>

namespace prime_factors {
const std::vector<long long> of(long long n) {
  auto result = std::vector<long long>();

  int i = 2;
  while (n > 1) {
    if (n % i == 0) {
      n /= i;
      result.push_back(i);
    } else {
      i++;
    }
  }

  return result;
}
} // namespace prime_factors

File: ocaml/prime-factors/prime_factors.ml
let factors_of n =
  let ( % ) = Int64.rem and ( + ) = Int64.add and ( / ) = Int64.div in
  let rec factors_of0 n divisor result =
    if n < 2L then result
    else if n % divisor = 0L then
      factors_of0 (n / divisor) divisor (divisor :: result)
    else factors_of0 n (divisor + 1L) result
  in
  factors_of0 n 2L [] |> List.rev
